src/treeutils/optimise.jl

```diff
@@ -0,0 +1,249 @@
+"""
+circuit optimisation
+"""
+module Optimise
+using SimpleTraits
+using YaoBlocks, YaoBlocks.ConstGate
+
+include("to_basictypes.jl")
+
+export replace_block
+"""
+    replace_block(actor, tree::AbstractBlock) -> AbstractBlock
+    replace_block(pair::Pair{Type{ST}, TT}, tree::AbstractBlock) -> AbstractBlock
+
+replace blocks in a circuit, where `actor` is a function that given input block,
+returns the block to replace, or `nothing` for skip replacing and visit sibling.
+If `pair` is provided, then replace original block with type `ST` with new block (`pair.second`).
+"""
+function replace_block(actor, tree::AbstractBlock)
+    res = actor(tree)
+    if res === tree || res isa Nothing # not replaced
```

Comment:
compiler prefers  `x === nothing` more

Reply:
Why? here chris also use `isa`: https://mitmath.github.io/18337/lecture2/optimizing

Reply:
Anyway, I have fixed it.